### ISSUE TITLE
add support for ENC data

### ISSUE THREAD

**Issue by flaf** [nwops/puppet-repl#12]
Hi,

In my puppetserver (Puppet 4), I have an ENC which defines the global variable `$::included_classes`. It's an array which contains all classes to apply  to the current node. If this array is empty or if the variable is undefined, the code in `site.pp` raises an error and stop the puppet run. It's a **wanted** feature in my `site.pp` code.

The consequence is that puppet-repl triggers an error at each command (even with an empty command):

```sh
[0] root@puppet.athome.priv 23:46 ~ # lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.4 LTS
Release:	14.04
Codename:	trusty

[0] root@puppet.athome.priv 23:47 ~ # prepl
Ruby Version: 2.1.8
Puppet Version: 4.3.2
Puppet Repl Version: 0.0.4
Created by: NWOps <corey@nwops.io>
Type "exit", "functions", "facts", "reset", "help" for more information.

>> a = 1
 => Evaluation Error: Error while evaluating a Function Call, Sorry, the node must have a
`$::included_classes` global variable defined (for instance by the ENC) and it must be an
non-empty array of elements with this form `[ '<author>', '<fully-qualified-class-name>' ]`
at /etc/puppetlabs/code/environments/production/manifests/site.pp:123:3
>> 
 => Evaluation Error: Error while evaluating a Function Call, Sorry, the node must have a
`$::included_classes` global variable defined (for instance by the ENC) and it must be an
non-empty array of elements with this form `[ '<author>', '<fully-qualified-class-name>' ]`
at /etc/puppetlabs/code/environments/production/manifests/site.pp:123:3
>> 
```

Anyway, thx for this promising tool.
Regards.

**Comment by logicminds** [comment]
At this time the repl does load the site.pp as its extremely helpful for setting topscope variables during development and playing with the language. 

In order to support ENC data the repl would have to call out to the ENC with a context (node info) in order to get the data from the ENC and set the top scope based on the returned data.  I would much rather leverage puppet to perform those actions than create some hooks to perform the same functionality. 

This is definitely something the repl will need to support going forward.

**Comment by logicminds** [comment]
@flaf do you have time to try the devel branch out?  I have fixed this in #14 but have only tested with a PE rig.  You would need to do the following.

````
gem build puppet-repl.gemspec
gem install puppet-repl-0.2.0.gem
prepl -n node_name
```

Note: this would reach out to the puppet server configured in your local puppet.conf and essentially run `puppet node find node_name --terminus rest`

The site.pp on the master does not evaluated since your local site.pp as defined in your puppet.conf is evaulated instead.  So it really only pulls in the remote node data and still evaluates everything locally.

You will also need to do the following to allow API access to the the node objects

See #17 for more info.

**Comment by flaf** [comment]
@logicminds, is it what you want?

```sh
root@puppet.athome.priv 01:14 ~# git clone https://github.com/nwops/puppet-repl.git
Cloning into 'puppet-repl'...
remote: Counting objects: 546, done.
remote: Compressing objects: 100% (121/121), done.
remote: Total 546 (delta 63), reused 0 (delta 0), pack-reused 421
Receiving objects: 100% (546/546), 905.41 KiB | 678.00 KiB/s, done.
Resolving deltas: 100% (298/298), done.
Checking connectivity... done.

root@puppet.athome.priv 01:16 ~ # cd puppet-repl/

root@puppet.athome.priv 01:16 ~/puppet-repl # git checkout devel 
Branch devel set up to track remote branch devel from origin.
Switched to a new branch 'devel'

root@puppet.athome.priv 01:17 ~/puppet-repl # gem build puppet-repl.gemspec
WARNING:  open-ended dependency on puppet (>= 3.8) is not recommended
  if puppet is semantically versioned, use:
    add_runtime_dependency 'puppet', '~> 3.8'
WARNING:  open-ended dependency on facterdb (>= 0) is not recommended
  if facterdb is semantically versioned, use:
    add_runtime_dependency 'facterdb', '~> 0'
WARNING:  open-ended dependency on awesome_print (>= 0) is not recommended
  if awesome_print is semantically versioned, use:
    add_runtime_dependency 'awesome_print', '~> 0'
WARNING:  open-ended dependency on pry (>= 0) is not recommended
  if pry is semantically versioned, use:
    add_runtime_dependency 'pry', '~> 0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: puppet-repl
  Version: 0.2.0
  File: puppet-repl-0.2.0.gem

root@puppet.athome.priv 01:17 ~/puppet-repl # gem install puppet-repl-0.2.0.gem
Fetching: awesome_print-1.6.1.gem (100%)
Successfully installed awesome_print-1.6.1
Fetching: coderay-1.1.1.gem (100%)
Successfully installed coderay-1.1.1
Fetching: method_source-0.8.2.gem (100%)
Successfully installed method_source-0.8.2
Fetching: slop-3.6.0.gem (100%)
Successfully installed slop-3.6.0
Fetching: pry-0.10.3.gem (100%)
Successfully installed pry-0.10.3
Successfully installed puppet-repl-0.2.0
Parsing documentation for awesome_print-1.6.1
Installing ri documentation for awesome_print-1.6.1
invalid options: -SNw2
(invalid options are ignored)
Parsing documentation for coderay-1.1.1
Installing ri documentation for coderay-1.1.1
Parsing documentation for method_source-0.8.2
Installing ri documentation for method_source-0.8.2
Parsing documentation for pry-0.10.3
Installing ri documentation for pry-0.10.3
Parsing documentation for puppet-repl-0.2.0
Installing ri documentation for puppet-repl-0.2.0
Parsing documentation for slop-3.6.0
Installing ri documentation for slop-3.6.0
Done installing documentation for awesome_print, coderay, method_source, pry, puppet-repl, slop after 7 seconds 6 gems installed

root@puppet.athome.priv 01:25 ~/puppet-repl # ./bin/prepl puppet.athome.priv
Ruby Version: 2.1.8
Puppet Version: 4.4.2
Puppet Repl Version: 0.2.0
Created by: NWOps <corey@nwops.io>
Type "exit", "functions", "vars", "krt", "facts", "resources", "classes",
     "play", "classification", "reset", or "help" for more information.

/root/puppet-repl/lib/puppet-repl/cli.rb:186:in `filename': No such file or directory @ rb_sysopen - puppet.athome.priv (Errno::ENOENT)
	from /root/puppet-repl/lib/puppet-repl/cli.rb:186:in `start'
	from ./bin/prepl:4:in `<main>'

root@puppet.athome.priv 01:26 ~/puppet-repl # puppet node find puppet.athome.priv --terminus rest
(works fine, I have big json in the output)
```

**Comment by logicminds** [comment]
Need to use the '-n'  `prepl -n node_name`

**Comment by logicminds** [comment]
So if that loads correctly you should be able to use

`classification` or `vars` or `facts` and anything else.

Also added support for variable and function name auto completion

**Comment by flaf** [comment]
Ooops, sorry:

```sh
root@puppet.athome.priv 01:43 ~/puppet-repl # ./bin/prepl -n puppet.athome.priv
Ruby Version: 2.1.8
Puppet Version: 4.4.2
Puppet Repl Version: 0.2.0
Created by: NWOps <corey@nwops.io>
Type "exit", "functions", "vars", "krt", "facts", "resources", "classes",
     "play", "classification", "reset", or "help" for more information.

>> a = 1
Fetching node puppet.athome.priv
/root/puppet-repl/lib/puppet-repl/support/scope.rb:29:in `rescue in create_scope': undefined method `parse_error' for #<PuppetRepl::Cli:0x000000042534c0> (NoMethodError)
	from /root/puppet-repl/lib/puppet-repl/support/scope.rb:26:in `create_scope'
	from /root/puppet-repl/lib/puppet-repl/support/scope.rb:12:in `scope'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:45:in `puppet_eval'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:111:in `handle_input'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:158:in `read_loop'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:196:in `start'
	from ./bin/prepl:4:in `<main>'
```

**Comment by logicminds** [comment]
1. Looks like you may not have changed your auth.conf  see #17 
2. Also I have an error in the code, otherwise you would have seen a better error message


Reproducing now and will push changes shortly

**Comment by flaf** [comment]
Here is my `/etc/puppetlabs/puppetserver/conf.d/auth.conf` after modification (I have replaced `"$1"` by `"*"`):

```sh
root@puppet.athome.priv 02:05 ~/puppet-repl
# cat /etc/puppetlabs/puppetserver/conf.d/auth.conf 
authorization: {
    version: 1
    rules: [
        {
            # Allow nodes to retrieve their own catalog
            match-request: {
                path: "^/puppet/v3/catalog/([^/]+)$"
                type: regex
                method: [get, post]
            }
            allow: "$1"
            sort-order: 500
            name: "puppetlabs catalog"
        },
        {
            # Allow nodes to retrieve the certificate they requested earlier
            match-request: {
                path: "/puppet-ca/v1/certificate/"
                type: path
                method: get
            }
            allow-unauthenticated: true
            sort-order: 500
            name: "puppetlabs certificate"
        },
        {
            # Allow all nodes to access the certificate revocation list
            match-request: {
                path: "/puppet-ca/v1/certificate_revocation_list/ca"
                type: path
                method: get
            }
            allow-unauthenticated: true
            sort-order: 500
            name: "puppetlabs crl"
        },
        {
            # Allow nodes to request a new certificate
            match-request: {
                path: "/puppet-ca/v1/certificate_request"
                type: path
                method: [get, put]
            }
            allow-unauthenticated: true
            sort-order: 500
            name: "puppetlabs csr"
        },
        {
            match-request: {
                path: "/puppet/v3/environments"
                type: path
                method: get
            }
            allow: "*"
            sort-order: 500
            name: "puppetlabs environments"
        },
        {
            # Allow nodes to access all file services; this is necessary for
            # pluginsync, file serving from modules, and file serving from
            # custom mount points (see fileserver.conf). Note that the `/file`
            # prefix matches requests to file_metadata, file_content, and
            # file_bucket_file paths.
            match-request: {
                path: "/puppet/v3/file"
                type: path
            }
            allow: "*"
            sort-order: 500
            name: "puppetlabs file"
        },
        {
            # Allow nodes to retrieve only their own node definition
            match-request: {
                path: "^/puppet/v3/node/([^/]+)$"
                type: regex
                method: get
            }
            allow: "*"
            sort-order: 500
            name: "puppetlabs node"
        },
        {
            # Allow nodes to store only their own reports
            match-request: {
                path: "^/puppet/v3/report/([^/]+)$"
                type: regex
                method: put
            }
            allow: "$1"
            sort-order: 500
            name: "puppetlabs report"
        },
        {
            match-request: {
                path: "/puppet/v3/status"
                type: path
                method: get
            }
            allow-unauthenticated: true
            sort-order: 500
            name: "puppetlabs status"
        },
        {
            match-request: {
                path: "/puppet/v3/static_file_content"
                type: path
                method: get
            }
            allow: "*"
            sort-order: 500
            name: "puppetlabs static file content"
        },
        {
          # Deny everything else. This ACL is not strictly
          # necessary, but illustrates the default policy
          match-request: {
            path: "/"
            type: path
          }
          deny: "*"
          sort-order: 999
          name: "puppetlabs deny all"
        }
    ]
}
```

I have restarted the puppetserver and:

```sh
root@puppet.athome.priv 02:05 ~/puppet-repl # ./bin/prepl -n puppet.athome.priv
Ruby Version: 2.1.8
Puppet Version: 4.4.2
Puppet Repl Version: 0.2.0
Created by: NWOps <corey@nwops.io>
Type "exit", "functions", "vars", "krt", "facts", "resources", "classes",
     "play", "classification", "reset", or "help" for more information.

>> a = 1
Fetching node puppet.athome.priv
/root/puppet-repl/lib/puppet-repl/support/scope.rb:29:in `rescue in create_scope': undefined method `parse_error' for #<PuppetRepl::Cli:0x00000003a56e20> (NoMethodError)
	from /root/puppet-repl/lib/puppet-repl/support/scope.rb:26:in `create_scope'
	from /root/puppet-repl/lib/puppet-repl/support/scope.rb:12:in `scope'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:45:in `puppet_eval'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:111:in `handle_input'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:158:in `read_loop'
	from /root/puppet-repl/lib/puppet-repl/cli.rb:196:in `start'
	from ./bin/prepl:4:in `<main>'
```

**Comment by logicminds** [comment]
Cool, I just pushed an update if you want to pull and rebuild the gem and try again.

**Comment by flaf** [comment]
```sh
root@puppet.athome.priv 02:12 ~/puppet-repl
# ./bin/prepl -n puppet.athome.priv
Ruby Version: 2.1.8
Puppet Version: 4.4.2
Puppet Repl Version: 0.2.0
Created by: NWOps <corey@nwops.io>
Type "exit", "functions", "vars", "krt", "facts", "resources", "classes",
     "play", "classification", "reset", or "help" for more information.

>> a = 1
Fetching node puppet.athome.priv
 => Evaluation Error: Operator '[]' is not applicable to an Undef Value. at /etc/puppetlabs/code/environments/production/manifests/site.pp:58:4
```

**Comment by logicminds** [comment]
`a = 1` is not valid puppet code.  Try anything else: `$a = '1'` or `classes`

**Comment by logicminds** [comment]
or `classification`

**Comment by flaf** [comment]
Oh sorry, of course `a = 1` is not valid. ;)

```sh
# ./bin/prepl -n puppet.athome.priv
Ruby Version: 2.1.8
Puppet Version: 4.4.2
Puppet Repl Version: 0.2.0
Created by: NWOps <corey@nwops.io>
Type "exit", "functions", "vars", "krt", "facts", "resources", "classes",
     "play", "classification", "reset", or "help" for more information.

>> $a = 'foo'
Fetching node puppet.athome.priv
 => Evaluation Error: Operator '[]' is not applicable to an Undef Value. at /etc/puppetlabs/code/environments/production/manifests/site.pp:58:4
>> classes
 => Evaluation Error: Operator '[]' is not applicable to an Undef Value. at /etc/puppetlabs/code/environments/production/manifests/site.pp:58:4
```

**Comment by logicminds** [comment]
Are you able to share your site.pp code?  Whats on line 58?

`/etc/puppetlabs/code/environments/production/manifests/site.pp:58:4`

**Comment by flaf** [comment]
It's `if $::facts['os']['family'].downcase == 'debian' {`

**Comment by logicminds** [comment]
Also what ENC are you using?

1. PE console
2. Foreman
3. Some script
4. hiera

**Comment by flaf** [comment]
My ENC is a personal Ruby script.

```
root@puppet.athome.priv 02:21 ~/puppet-repl # /etc/puppetlabs/code/environments/enc puppet.athome.priv
{
  "environment": "production",
  "parameters":
  {
    "included_classes": [["flaf", "::roles::puppetserver"]],
    "datacenter": "vboxhome",
    "groups": ["puppetserver"],
    "groups_0": "puppetserver",
    "groups_1": "NOT_AVAILABLE",
    "groups_2": "NOT_AVAILABLE",
    "groups_3": "NOT_AVAILABLE",
    "groups_4": "NOT_AVAILABLE",
    "datacenters": ["vboxhome"],
  }
}
```

**Comment by logicminds** [comment]
Just pushed a new update.  Can you pull, rebuild and try again?

There is an issue with the puppet node face not populating the $::facts variable so I retrofitted it in.

**Comment by flaf** [comment]
Hi,

During the command `gem install puppet-repl-0.2.0.gem`, some files in `/opt/puppetlabs/puppet/bin/` are overwritten. Is it normal? No problem here because it's a VM with a snapshot.

```sh
[0] root@puppet.atwork.priv 15:38 ~/puppet-repl
# gem install puppet-repl-0.2.0.gem
Fetching: facter-2.4.6.gem (100%)
facter's executable "facter" conflicts with /opt/puppetlabs/puppet/bin/facter
Overwrite the executable? [yN]  y
Successfully installed facter-2.4.6
Fetching: json_pure-1.8.3.gem (100%)
Successfully installed json_pure-1.8.3
Fetching: hiera-1.3.4.gem (100%)
hiera's executable "hiera" conflicts with /opt/puppetlabs/puppet/bin/hiera
Overwrite the executable? [yN]  y
Successfully installed hiera-1.3.4
Fetching: puppet-3.8.7.gem (100%)
puppet's executable "puppet" conflicts with /opt/puppetlabs/puppet/bin/puppet
Overwrite the executable? [yN]  y
Successfully installed puppet-3.8.7
Fetching: jgrep-1.4.0.gem (100%)
Successfully installed jgrep-1.4.0
Fetching: facterdb-0.3.5.gem (100%)
Successfully installed facterdb-0.3.5
Fetching: awesome_print-1.6.1.gem (100%)
Successfully installed awesome_print-1.6.1
Successfully installed puppet-repl-0.2.0
Parsing documentation for awesome_print-1.6.1
Installing ri documentation for awesome_print-1.6.1
Parsing documentation for facter-2.4.6
Installing ri documentation for facter-2.4.6
Parsing documentation for facterdb-0.3.5
Installing ri documentation for facterdb-0.3.5
Parsing documentation for hiera-1.3.4
Installing ri documentation for hiera-1.3.4
Parsing documentation for jgrep-1.4.0
Installing ri documentation for jgrep-1.4.0
Parsing documentation for json_pure-1.8.3
Installing ri documentation for json_pure-1.8.3
Parsing documentation for puppet-3.8.7
Installing ri documentation for puppet-3.8.7
Parsing documentation for puppet-repl-0.2.0
Installing ri documentation for puppet-repl-0.2.0
Done installing documentation for awesome_print, facter, facterdb, hiera, jgrep, json_pure, puppet, puppet-repl after 37 seconds
8 gems installed
```

And now I have this:

```sh
[0] root@puppet.atwork.priv 15:46 ~/puppet-repl
# ./bin/prepl -n puppet.atwork.priv
Ruby Version: 2.1.8
Puppet Version: 4.4.2
Puppet Repl Version: 0.2.0
Created by: NWOps <corey@nwops.io>
Type "exit", "functions", "vars", "krt", "facts", "resources", "classes",
     "play", "classification", "reset", or "help" for more information.

>> $a = '1'
Fetching node puppet.atwork.priv
 => Evaluation Error: Unknown variable: '::server_facts'. at /etc/puppetlabs/code/environments/production/modules/puppetagent/functions/data.pp:6:6 in Evaluation Error: Unknown variable: '::server_facts'. at /etc/puppetlabs/code/environments/production/modules/puppetagent/functions/data.pp:6:6
>> exit
```

Here is the line of my `/etc/puppetlabs/code/environments/production/modules/puppetagent/functions/data.pp`:

```puppet
function puppetagent::data {

  # Warning: the $server_facts will be defined for the node
  #          only if the parameter `trusted_server_facts`
  #          is set to true in the puppet.conf of the server.
  if $::server_facts {              #  <==== line 6 here

  ...
```

**Comment by logicminds** [comment]
@flaf I have updated the branch (force push) if you want to delete the repo and try again. 

I ended up having to duplicate the server_facts on the "client side" as the server does not expose those facts to via the puppet node indirector.  

Example: https://www.puppet-repl.com/play?content=vars

**Comment by logicminds** [comment]
released with v0.2.0
